### PR TITLE
Upgrades

### PR DESCRIPTION
**Pull request by jbr**
this pr is really just the last commit

### CODE REVIEW

**Comment by jbr** [comment]
ci failing because it doesn't have a crates-io patch for http-types

File: src/server/mod.rs [hunk 1]
```diff
@@ -70,14 +71,32 @@ where
             }
         };
 
+        let upgrade_requested = match (req.header(UPGRADE), req.header(CONNECTION)) {
+            (Some(_), Some(upgrade)) if upgrade.as_str().eq_ignore_ascii_case("upgrade") => true,
```
Comment (Fishrock123) [inline src/server/mod.rs:75]: `Some(upgrade)` would probably be clear as `Some(connection)`

File: src/server/mod.rs [hunk 2]
```diff
@@ -70,14 +71,32 @@ where
...
+            _ => false,
+        };
+
         let method = req.method();
+
         // Pass the request to the endpoint and encode the response.
-        let res = endpoint(req).await?;
+        let mut res = endpoint(req).await?;
+
+        let upgrade_provided = res.status() == StatusCode::SwitchingProtocols && res.has_upgrade();
```
Comment (Fishrock123) [inline src/server/mod.rs:84]: is `res.has_upgrade()` done automatically for default endpoints?

Comment (jbr) [reply src/server/mod.rs:84]: yes, it defaults to false — that's already on http_types::Response

**Review by yoshuawuyts — APPROVED** [review]
The changes in the last commit look good!

**Pull request merged**